src/Microsoft.VisualStudio.Mac.LanguageServices.Razor/ProjectSystem/DefaultDotNetProjectHost.cs

```diff
@@ -0,0 +1,135 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.VisualStudio.Editor.Razor;
+using MonoDevelop.Projects;
+
+namespace Microsoft.VisualStudio.Mac.LanguageServices.Razor.ProjectSystem
+{
+    internal class DefaultDotNetProjectHost : DotNetProjectHost
+    {
+        private const string ExplicitRazorConfigurationCapability = "DotNetCoreRazorConfiguration";
+
+        private readonly DotNetProject _project;
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+        private readonly VisualStudioMacWorkspaceAccessor _workspaceAccessor;
+        private readonly TextBufferProjectService _projectService;
+        private RazorProjectHostBase _razorProjectHost;
+
+        public DefaultDotNetProjectHost(
+            DotNetProject project,
+            ForegroundDispatcher foregroundDispatcher,
+            VisualStudioMacWorkspaceAccessor workspaceAccessor,
+            TextBufferProjectService projectService)
+        {
+            if (project == null)
+            {
+                throw new ArgumentNullException(nameof(project));
+            }
+
+            if (foregroundDispatcher == null)
+            {
+                throw new ArgumentNullException(nameof(foregroundDispatcher));
+            }
+
+            if (workspaceAccessor == null)
+            {
+                throw new ArgumentNullException(nameof(workspaceAccessor));
+            }
+
+            if (projectService == null)
+            {
+                throw new ArgumentNullException(nameof(projectService));
+            }
+
+            _project = project;
+            _foregroundDispatcher = foregroundDispatcher;
+            _workspaceAccessor = workspaceAccessor;
+            _projectService = projectService;
+        }
+
+        public override DotNetProject Project => _project;
+
+        public override void Subscribe()
+        {
+            _foregroundDispatcher.AssertForegroundThread();
+
+            UpdateRazorHostProject();
+
+            _project.ProjectCapabilitiesChanged += Project_ProjectCapabilitiesChanged;
+            _project.Disposing += Project_Disposing;
+        }
+
+        private void Project_Disposing(object sender, EventArgs e)
+        {
+            _foregroundDispatcher.AssertForegroundThread();
+
+            _project.ProjectCapabilitiesChanged -= Project_ProjectCapabilitiesChanged;
+            _project.Disposing -= Project_Disposing;
+
+            DetatchCurrentRazorProjectHost();
+        }
+
+        private void Project_ProjectCapabilitiesChanged(object sender, EventArgs e) => UpdateRazorHostProject();
+
+        private void UpdateRazorHostProject()
+        {
+            _foregroundDispatcher.AssertForegroundThread();
+
+            DetatchCurrentRazorProjectHost();
```

Comment:
Just a thought here, this seems like it will result in the project getting removed and re-created. That's not super desirable.

Reply:
You're correct it will. It'll only ever happen when the capabilities of the project change though. We could be more conditional about re-creating the project based on the capabilities if we really want but the capabilities changing happens so infrequently in VS4Mac world.